controllers/tlspolicy_certmanager_certificates.go

```diff
@@ -130,12 +131,36 @@ func (r *TLSPolicyReconciler) expectedCertificatesForGateway(ctx context.Context
 
 	certs := make([]*certmanv1.Certificate, 0, len(tlsHosts))
 	for secretRef, hosts := range tlsHosts {
-		certs = append(certs, r.buildCertManagerCertificate(gateway, tlsPolicy, secretRef, hosts))
+		certs = append(certs, buildCertManagerCertificate(gateway, tlsPolicy, secretRef, hosts))
 	}
 	return certs
 }
 
-func (r *TLSPolicyReconciler) buildCertManagerCertificate(gateway *gatewayapiv1.Gateway, tlsPolicy *v1alpha1.TLSPolicy, secretRef corev1.ObjectReference, hosts []string) *certmanv1.Certificate {
+func expectedCertificatesForListener(l *machinery.Listener, tlsPolicy *v1alpha1.TLSPolicy) []*certmanv1.Certificate {
+	tlsHosts := make(map[corev1.ObjectReference][]string)
+
+	for _, certRef := range l.TLS.CertificateRefs {
+		secretRef := corev1.ObjectReference{
+			Name: string(certRef.Name),
+		}
+		if certRef.Namespace != nil {
+			secretRef.Namespace = string(*certRef.Namespace)
+		} else {
+			secretRef.Namespace = l.GetNamespace()
+		}
+		// Gateway API hostname explicitly disallows IP addresses, so this
+		// should be OK.
+		tlsHosts[secretRef] = append(tlsHosts[secretRef], string(*l.Hostname))
```

Comment:
I believe if the listener has no hostname, this line will throw a nil pointer dereference exception.

Reply:
A workaround (backed by the [docs](https://pkg.go.dev/sigs.k8s.io/gateway-api@v1.1.0/apis/v1#Listener.Hostname)) is considering `*` when `l.Hostname == nil`.